app/services/idv/gpo_mail.rb

```diff
@@ -39,19 +41,47 @@ def hours_since_first_letter(first_letter_requested_at)
 
     attr_reader :current_user
 
-    def user_mail_events
-      @user_mail_events ||= current_user.events.
-        gpo_mail_sent.
-        order('updated_at DESC').
-        limit(MAX_MAIL_EVENTS)
+    def rate_limiting_enabled?
+      window_limit_enabled? || last_not_too_recent_enabled?
     end
 
-    def max_events?
-      user_mail_events.size == MAX_MAIL_EVENTS
+    def window_limit_enabled?
+      MAX_MAIL_EVENTS != 0 && MAIL_EVENTS_WINDOW_DAYS != 0
     end
 
-    def updated_within_last_month?
-      user_mail_events.last.updated_at > MAIL_EVENTS_WINDOW_DAYS.days.ago
+    def last_not_too_recent_enabled?
+      MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER_IN_HOURS != 0 &&
+        current_user.pending_profile?
+    end
+
+    def too_many_mails_within_window?
+      return false unless window_limit_enabled?
+
+      number_of_mails_within(
+        MAIL_EVENTS_WINDOW_DAYS.days,
+      ) >= MAX_MAIL_EVENTS
+    end
+
+    def last_mail_too_recent?
+      return false unless last_not_too_recent_enabled?
+
+      number_of_mails_within(
+        MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER_IN_HOURS.hours,
+        for_profile: current_user.pending_profile,
+      ) > 0
+    end
+
+    def number_of_mails_within(time_window, for_profile: nil)
+      profile_hash = { user: current_user }
+      profile_hash[:id] = for_profile.id if for_profile
+
+      GpoConfirmationCode.joins(:profile).
+        where(
+          updated_at: (time_window.ago..),
+          profile: profile_hash,
+        ).
+        limit([1, MAX_MAIL_EVENTS].max).
+        count
```

Comment:
Could do a single query that sorts by timestamp, and then compare the most recent timestamp against the minimum wait in hours, rather than doing two similar queries.

Reply:
Hmm. Will think about this. My first reaction is that it doesn't matter; asking for a count of one or zero events within the last 24 hours doesn't strike me as inherently more expensive than asking for a sort limited to one and retrieving the actual timestamp. The query as written can work entirely with indices, which is nice.

Reply:
See my response to @mitchellhenke's query comment; I left a small bit of debug code in place there. Removed.